paprika/restraints/plumed.py

```diff
@@ -0,0 +1,348 @@
+                # Skip restraint if the target or force constant is not defined.
+                # Example: wall restraints only used during the attach phase.
+                try:
+                    target = restraint.phase[phase]["targets"][window]
+                    force_constant = (
+                        restraint.phase[phase]["force_constants"][window]
+                        * self.k_factor
+                    )
+                except TypeError:
+                    continue
+
+                # Convert list to comma-separated string
+                atom_index = self._get_atom_indices(restraint)
+                atom_string = ",".join(map(str, atom_index))
+
+                # Determine collective variable type
+                colvar_type = "distance"
+                if len(atom_index) == 3:
+                    colvar_type = "angle"
+                    target *= _PI_ / 180.0
+                elif len(atom_index) == 4:
+                    colvar_type = "torsion"
+                    target *= _PI_ / 180.0
+
+                # Determine bias type for this restraint
+                bias_type = get_bias_potential_type(restraint, phase, window)
+                if bias_type is "harmonic":
+                    bias_type = "restraint"
+
+                # Append string to lists
+                cv_list.append(
+                    f"c{cv_index}: {colvar_type.upper()} ATOMS={atom_string} NOPBC\n"
+                )
+                bias_list.append(
+                    f"{bias_type.upper()} ARG=c{cv_index} AT={target:.4f} KAPPA={force_constant:.2f}\n"
+                )
+
+                # Increment colvar index
+                cv_index += 1
+
+            # Write collective variables to file
+            self._write_colvar_to_file(windows, cv_list, bias_list)
+
+    def _write_colvar_to_file(self, window, cv_list, bias_list):
+        with open(os.path.join(self.path, window, self.file_name), "a") as file:
+            if len(self.group_atoms) != 0:
+                file.write("# Center groups\n")
+                for key, value in self.group_atoms.items():
+                    file.write(f"{key}: COM ATOMS={value}\n")
+
+            file.write("# Collective variables\n")
+            for line in cv_list:
+                file.write(line)
+
+            file.write("# Bias potentials\n")
+            for line in bias_list:
+                file.write(line)
+
+    def _get_atom_indices(self, restraint):
+        # Check atom index setting
+        index_shift = 0
+        if not restraint.amber_index:
+            index_shift = 1
+            logger.info("Atom indices starts from 0 --> shifting indices by 1.")
+
+        # Collect DAT atom indices
+        atom_index = []
+        for i in range(4):
+            ii = i + 1
+
+            if not eval(f"restraint.index{ii}"):
+                if ii in [1, 2]:
+                    raise Exception("There must be at least two atoms in a restraint.")
```

Comment:
Presumably before this point we've already tested for that, right? So this might be unnecessary?

Reply:
I don't think it has been tested before reaching this code. Also, I don't think `check_restraints` checks if there are at least two atoms in a restraint. https://github.com/slochower/pAPRika/blob/96219fea95ced11f9785857bdfda665ea1acc8ae/paprika/restraints/restraints.py#L856

Reply:
Ah okay.  I'd suggest putting a tiny function in `restraints/restraints.py` that checks for at least two atoms in a restraint. I think it'll be a little cleaner, plus we should call that anyway outside of PLUMED stuff.

Reply:
Done, the checking is done in `check_restraints` and I've updated the code for `amber_restraint_line` as well.